### PR TITLE
Add ceil_mode to max pooling and average pooling

### PR DESCRIPTION
**Pull request by hzxie**
See PR [#3057 in Caffe](https://github.com/BVLC/caffe/pull/3057).

**However**, I cannot build tests and examples.
I think it is caused  by something defined in `caffe.proto`.

I've added following line into `caffe.proto`, but still not working.
```
optional bool ceil_mode = 13 [default = false];
```

And here's the error message:
```
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/x86_64-pc-linux-gnu/bits/c++allocator.h:33:0,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/bits/allocator.h:46,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/string:41,
                 from /home/hzxie/Development/tiny-dnn/test/cotire/tiny_dnn_test_CXX_prefix.cxx:6,
                 from /home/hzxie/Development/tiny-dnn/test/cotire/tiny_dnn_test_CXX_prefix.hxx:4:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = tiny_dnn::max_pooling_layer; _Args = {const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&}; _Tp = tiny_dnn::max_pooling_layer]’:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/bits/alloc_traits.h:475:4:   required from ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = tiny_dnn::max_pooling_layer; _Args = {const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&}; _Tp = tiny_dnn::max_pooling_layer; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<tiny_dnn::max_pooling_layer>]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/bits/shared_ptr_base.h:526:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&}; _Tp = tiny_dnn::max_pooling_layer; _Alloc = std::allocator<tiny_dnn::max_pooling_layer>; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/bits/shared_ptr_base.h:637:4:   required from ‘std::__shared_count<_Lp>::__shared_count(std::_Sp_make_shared_tag, _Tp*, const _Alloc&, _Args&& ...) [with _Tp = tiny_dnn::max_pooling_layer; _Alloc = std::allocator<tiny_dnn::max_pooling_layer>; _Args = {const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&}; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/bits/shared_ptr_base.h:1295:35:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<tiny_dnn::max_pooling_layer>; _Args = {const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&}; _Tp = tiny_dnn::max_pooling_layer; __gnu_cxx::_Lock_policy _Lp = (__gnu_cxx::_Lock_policy)2]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/bits/shared_ptr.h:344:64:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_make_shared_tag, const _Alloc&, _Args&& ...) [with _Alloc = std::allocator<tiny_dnn::max_pooling_layer>; _Args = {const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&}; _Tp = tiny_dnn::max_pooling_layer]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/bits/shared_ptr.h:690:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = tiny_dnn::max_pooling_layer; _Alloc = std::allocator<tiny_dnn::max_pooling_layer>; _Args = {const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&}]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/bits/shared_ptr.h:706:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = tiny_dnn::max_pooling_layer; _Args = {const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&}]’
/home/hzxie/Development/tiny-dnn/tiny_dnn/io/caffe/layer_factory_impl.h:139:46:   required from here
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.1/include/c++/ext/new_allocator.h:136:4: error: no matching function for call to ‘tiny_dnn::max_pooling_layer::max_pooling_layer(const long unsigned int&, const long unsigned int&, const long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, long unsigned int&, tiny_dnn::padding&)’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
```

What should I do to fix this error?

### CODE REVIEW

**Comment by hzxie** [comment]
The compile error can be resolved by adding two constructors (that makes API unchanged):

```
max_pooling_layer(size_t in_width,
                    size_t in_height,
                    size_t in_channels,
                    size_t pooling_size_x,
                    size_t pooling_size_y,
                    size_t stride_x,
                    size_t stride_y,
                    padding pad_type             = padding::valid,
                    core::backend_t backend_type = core::default_engine())
    : max_pooling_layer(in_width,
                        in_height,
                        in_channels,
                        pooling_size_x,
                        pooling_size_y,
                        stride_x,
                        stride_y,
                        false,
                        padding::valid,
                        backend_type) {}
```

and 

```
average_pooling_layer(size_t in_width,
                        size_t in_height,
                        size_t in_channels,
                        size_t pool_size_x,
                        size_t pool_size_y,
                        size_t stride_x,
                        size_t stride_y,
                        padding pad_type = padding::valid) 
    : average_pooling_layer(in_width,
                            in_height,
                            in_channels,
                            pool_size_x,
                            pool_size_y,
                            stride_x,
                            stride_y,
                            false,
                            padding::valid) {}
```

But I don't think it is a good solution.

**Comment by hzxie** [comment]
@edgarriba  Please also review this PR.

File: tiny_dnn/layers/average_pooling_layer.h [hunk 1]
```diff
@@ -141,19 +150,41 @@ class average_pooling_layer : public partial_connected_layer {
...
+                            ceil_mode,
+                            padding::valid) {}
+
+  average_pooling_layer(size_t in_width,
+                        size_t in_height,
+                        size_t in_channels,
+                        size_t pool_size_x,
+                        size_t pool_size_y,
+                        size_t stride_x,
+                        size_t stride_y,
```
Comment (beru) [inline tiny_dnn/layers/average_pooling_layer.h]: I'm not really sure why this constructor does not have `ceil_mode` argument.
Would it be a problem?

File: tiny_dnn/layers/max_pooling_layer.h [hunk 2]
```diff
@@ -71,6 +78,27 @@ class max_pooling_layer : public layer {
...
+                        backend_type) {}
+
+  max_pooling_layer(size_t in_width,
+                    size_t in_height,
+                    size_t in_channels,
+                    size_t pooling_size_x,
+                    size_t pooling_size_y,
+                    size_t stride_x,
+                    size_t stride_y,
+                    padding pad_type             = padding::valid,
```
Comment (beru) [inline tiny_dnn/layers/max_pooling_layer.h]: The same question. https://github.com/tiny-dnn/tiny-dnn/pull/995/files#r226743476
It looks like only this constructor doesn't have `ceil_mode` parameter.

**Comment by beru** [comment]
@hzxie I didn't understand the problem. Let me check how to update `tiny_dnn/io/caffe/layer_factory_impl.h`.

However, I still think this library's design around having massive amount of overloaded constructors hinders maintainability...

**Comment by hzxie** [comment]
@beru 
The problem has been resolved by adding corresponding parameters in `tiny_dnn/io/caffe/layer_factory_impl.h`.
Please review it again.

File: tiny_dnn/util/util.h [hunk 3]
```diff
@@ -343,6 +343,28 @@ inline size_t conv_out_length(size_t in_length,
...
+    output_length = in_length - window_size + 1;
+  } else {
+    throw nn_error("Not recognized pad_type.");
+  }
+
+  if ( ceil_mode ) {
+    return static_cast<int>(ceil(static_cast<float>((output_length + stride - 1)) / stride));
+  } else {
+    return static_cast<int>(floor(static_cast<float>((output_length + stride - 1)) / stride));
+  }
```
Comment (beru) [inline tiny_dnn/util/util.h]: How about writing code in this way?
```cpp
  auto tmp = static_cast<float_t>(output_length + stride - 1) / stride;
  if ( ceil_mode ) {
    return static_cast<int>(ceil(tmp));
  } else {
    return static_cast<int>(floor(tmp));
  }
```

File: tiny_dnn/util/util.h [hunk 4]
```diff
@@ -358,10 +358,11 @@ inline size_t pool_out_length(size_t in_length,
...
   }
 
-  if ( ceil_mode ) {
-    return static_cast<int>(ceil(static_cast<float>((output_length + stride - 1)) / stride));
+  float tmp = static_cast<float>((output_length + stride - 1)) / stride;
+  if (ceil_mode) {
+    return static_cast<int>(ceil(tmp));
   } else {
-    return static_cast<int>(floor(static_cast<float>((output_length + stride - 1)) / stride));
+    return static_cast<int>(floor(tmp));
```
Comment (beru) [inline tiny_dnn/util/util.h]: Thank you for updating the code.
I've just thought there's another way of writing this code.

```cpp
  auto tmp = static_cast<float_t>(output_length + stride - 1) / stride;
  return static_cast<int>(ceil_mode ? ceil(tmp) : floor(tmp));
```

Comment (hzxie) [reply tiny_dnn/util/util.h]: Cool. Very simple.
I'll update it.

**Pull request merged**